programs/timefn.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2016-present, Yann Collet, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under both the BSD-style license (found in the
+ * LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ * in the COPYING file in the root directory of this source tree).
+ * You may select, at your option, one of the above-listed licenses.
+ */
+
+#ifndef TIME_FN_H_MODULE_287987
+#define TIME_FN_H_MODULE_287987
+
+#if defined (__cplusplus)
+extern "C" {
+#endif
+
+
+/*-****************************************
+*  Dependencies
+******************************************/
+#include <sys/types.h>    /* utime */
+#if defined(_MSC_VER)
+#  include <sys/utime.h>  /* utime */
+#else
+#  include <utime.h>      /* utime */
+#endif
+#include <time.h>         /* clock_t, clock, CLOCKS_PER_SEC */
+
+
+
+/*-****************************************
+*  Local Types
+******************************************/
+
+#if !defined (__VMS) && (defined (__cplusplus) || (defined (__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L) /* C99 */) )
+# include <stdint.h>
+  typedef uint64_t           PTime;  /* Precise Time */
+#else
+  typedef unsigned long long PTime;  /* does not support compilers without long long support */
+#endif
+
+
+
+/*-****************************************
+*  Time functions
+******************************************/
+#if defined(_WIN32)   /* Windows */
+
+    #include <Windows.h>   /* LARGE_INTEGER */
+    typedef LARGE_INTEGER UTIL_time_t;
+    #define UTIL_TIME_INITIALIZER { { 0, 0 } }
+
+#elif defined(__APPLE__) && defined(__MACH__)
+
+    #include <mach/mach_time.h>
+    typedef PTime UTIL_time_t;
+    #define UTIL_TIME_INITIALIZER 0
+
+#elif (defined (__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L) /* C11 */) \
```

Comment:
@Cyan4973 this is not the right feature test macro. This disables precise timing on all Linux machines not compiling with C11.

It used to be:

```
(PLATFORM_POSIX_VERSION >= 200112L) \	
   && (defined(__UCLIBC__)                \	
      || (defined(__GLIBC__)              \	
          && ((__GLIBC__ == 2 && __GLIBC_MINOR__ >= 17) \	
             || (__GLIBC__ > 2))))
```

Reply:
Also, this was detected by `poolTest` failing, so I guess that we're not currently running it in any CI.

Reply:
Right,
`PLATFORM_POSIX_VERSION` was a fairly complex macro, set within `platform.h`,
the result of mixing several other feature test macros and setting some build macros
in some surgical order, with plenty of exceptions.

In contrast, the `C11` test, while different, has the benefit of simplicity, and (at least in theory) a larger scope than posix.

The idea is to rely on `C90`'s `clock_t` when this condition could not be met.

Note that, whatever the code path, the number of nature of symbols defined remains constant. Same functions, same types, same macros.
`UTIL_clockSpanNano()` and `UTIL_getSpanTimeNano()` are still supposed to be supported, even with `C90`'s `clock_t`.

Reply:
The downside is that `clock_t` has very low resolution on Linux. Since this function is used in benchmarking, we'll end up with biased results.

Reply:
Yes.
This felt acceptable.
The design of the benchmark function should reduce the noise to ~1% on "low" resolution timers (10 ms).

A more problematic issue is that multi-thread benchmarks do not work with `clock_t` : it sums the time spent on each core. That's a more concerning limitation.

To be complete, I'm not against having a "posix branch", the same way there is already a Windows and a Mac OS-X branch.
But the way "posix compatibility" was tested looked a bit too scary for me, and I was afraid of its side-effects and maintenance.

All modern posix systems should be `C11` compatible by now, hence will correctly use the new `C11` path with high accuracy timers. 

So now it's a matter of improving the benchmark accuracy of old posix systems. It did not felt a high priority, but it can certainly be done if considered important enough. At least, with the `C11` path now available, the "posix" path only role is to improve timer accuracy for some targeted older systems. This might help to reduce the complexity of the macro test, which was staggering, as it used to be the only way to provide high-accuracy timers to linux and co, and was meant to pass compilation on all sort of unixes, many of which end up being "partially" compliant.